addons/binding/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/internal/UPBReader.java

```diff
@@ -0,0 +1,191 @@
+    }
+
+    /**
+     * Removes the listener from further messages.
+     *
+     * @param listener
+     *            the listener to remove.
+     */
+    public synchronized void removeListener(Listener listener) {
+        listeners.remove(listener);
+    }
+
+    /**
+     * Adds data to the buffer.
+     *
+     * @param data
+     *            the data to add.
+     * @param length
+     *            the length of data to add.
+     */
+    private void addData(byte[] data, int length) {
+
+        if (bufferLength + length > buffer.length) {
+            // buffer overflow discard entire buffer
+            bufferLength = 0;
+        }
+
+        System.arraycopy(data, 0, buffer, bufferLength, length);
+
+        bufferLength += length;
+
+        interpretBuffer();
+    }
+
+    /**
+     * Shuts the reader down.
+     */
+    public void shutdown() {
+        if (thread != null) {
+            thread.interrupt();
+        }
+
+        try {
+            inputStream.close();
+        } catch (IOException e) {
+        }
+    }
+
+    private int findMessageLength(byte[] buffer, int bufferLength) {
+        int messageLength = ArrayUtils.INDEX_NOT_FOUND;
+
+        for (int i = 0; i < bufferLength; i++) {
+            if (buffer[i] == 13) {
+                messageLength = i;
+                break;
+            }
+        }
+
+        return messageLength;
+    }
+
+    /**
+     * Attempts to interpret any messages that may be contained in the buffer.
+     */
+    private void interpretBuffer() {
+        int messageLength = findMessageLength(buffer, bufferLength);
+
+        while (messageLength != ArrayUtils.INDEX_NOT_FOUND) {
+            String message = new String(Arrays.copyOfRange(buffer, 0, messageLength));
+            logger.debug("UPB Message: {}", message);
+
+            int remainingBuffer = bufferLength - messageLength - 1;
+
+            if (remainingBuffer > 0) {
+                System.arraycopy(buffer, messageLength + 1, buffer, 0, remainingBuffer);
+            }
+            bufferLength = remainingBuffer;
+
+            notifyListeners(UPBMessage.fromString(message));
+
+            messageLength = findMessageLength(buffer, bufferLength);
+        }
+    }
+
+    private synchronized void notifyListeners(UPBMessage message) {
+        for (Listener l : new ArrayList<>(listeners)) {
+            l.messageReceived(message);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void run() {
+        logger.debug("Starting modem monitoring.");
+        byte[] buffer = new byte[32];
+        try {
+            for (int len = -1; (len = inputStream.read(buffer)) >= 0;) {
+                if (len > 0) {
+                    logger.debug("Received: {}", ArrayUtils.subarray(buffer, 0, len));
+                }
+                addData(buffer, len);
+                if (Thread.interrupted()) {
+                    break;
+                }
+            }
+        } catch (Exception e) {
```

Comment:
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception